pipeline/psth.py

```diff
@@ -19,129 +22,229 @@
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'hit',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_left_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'left'}
+            },
+            {
+                'trial_condition_desc': 'good_noearlylick_right_miss',
+                'trial_condition_func': '_get_trials_no_stim',
+                'trial_condition_arg': {
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'outcome': 'miss',
+                    'early_lick': 'no early',
+                    'trial_instruction': 'right'}
+            },
+        )
+
+        return ({**d, 'trial_condition_id': key_hash(d['trial_condition_arg'])}
+                for d in contents_data)
+
+    @classmethod
+    def get_trials(cls, trial_condition_desc):
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
+
+    @classmethod
+    def _get_trials_no_stim(cls, task=None, task_protocol=None, outcome=None,
+                            early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_no_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) - experiment.PhotostimEvent)
+
+    @classmethod
+    def _get_trials_stim(cls, task=None, task_protocol=None, outcome=None,
+                         early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) & experiment.PhotostimEvent)
+
+
+@schema
+class UnitPsth(dj.Computed):
+    definition = """
+    -> TrialCondition
+    -> ephys.Unit
+    ---
+    unit_psth=NULL:                             longblob
+    """
     psth_params = {'xmin': -3, 'xmax': 3, 'binsize': 0.04}
 
-    @staticmethod
-    def compute_unit_trial_psth(unit_key, trial_keys):
+    key_source = (ephys.Unit * TrialCondition
```

Comment:
Why are we restricting to the specific conditions? Just create relevant trial conditions and let this populate fully.

Reply:
There is one condition for all units w/o L/R which isn't needed for the plots; concievably (but not yet actually) possibly others; was initially added due to the 'handling None' issue discussed in other comment, but these actually are different issues (TrialCondition logic should handle None's, but this still doesn't mean all conditions will need UnitPSTH)..

can discuss further

Reply:
actually some other conditions are needed for plots (e.g. `plot_unit_bilateral_photostim_effect`, will adjust `Null` handling & add appropriate supporting criteria)